### PR TITLE
MM-46275/MM-46979 Production MPA for web app

### PR DESCRIPTION
**Pull request by hmhealey**
This PR is broken into two parts:
1. First, I made some changes to the web app Makefile and build scripts so that it's consistent with the server. Now, both repos will generate a tarball if you run `make dist` and the web app no longer has 3 different Makefile targets that do similar versions of that.
2. Second, I added the `setup-focalboard-product` CircleCI task to compile the Boards product files for the web app. It has a TODO since I ran into some issues with its `make build-product` task failing for some reason that I couldn't figure out. It just bypasses that now.

    That commit also includes changes to `make package` to package the files generated by `setup-focalboard-product` into the server tarball. Currently, it doesn't error out if those files don't exist, but we might also want to consider having it just error out if those files don't exist since I don't know if we intend on having people run `make package` locally.

One other thing I discussed that didn't end up needing any code changes was where I put the static files for the products in the final tarball. I thought I was going to add a new folder for it and serve those from a new API, but it turned out that I could just put them in `client/products` and have them automatically get served up by the existing code that handles web app static files.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-46275
https://mattermost.atlassian.net/browse/MM-46979

#### Related Pull Requests
https://github.com/mattermost/mattermost-webapp/pull/11355

#### Release Note
```release-note
Add logic to package product version of Boards with production builds
```

### CODE REVIEW

**Comment by agnivade** [comment]
> but we might also want to consider having it just error out if those files don't exist since I don't know if we intend on having people run make package locally.

The load tests require devs to run `make package` locally. But even then, we'd want to package the focalboard files as well. So I'd lean towards throwing an error.

File: build/release.mk
Comment (agnivade) [inline build/release.mk:154]: Shouldn't this be `/client/products`?

File: build/release.mk
Comment (agnivade) [inline build/release.mk:200]: Any need to run `ls`?

**Review by agnivade — COMMENTED** [review]
Tested out the steps and works fine. The only comment from my side would be to preserve the "make build" workflow for the client side.

Since I majorly work on the server, I rarely run the webapp. What I do is just do a `make build` on the webapp once, generate the files and keep them there. And just do the dev flow on the server.

With this change, I need to compile webapp, compile focalboard, AND the key part is to move over the focalboard assets to the right directory. We are doing this step only during the tarball creation job, but it would be beneficial to do it in the webapp `make build` phase itself.

But it's not a huge blocker since it's a custom workflow that I guess very few developers use and I would be fine in doing 2 more steps manually. I'll leave it to you.

File: .circleci/config.yml
Comment (phoinixgrr) [inline .circleci/config.yml:83]: Do we need  8 CPU Cores, and 16GB of RAM for this? 
If not let's lower this to something like `large` or `medium` 

https://circleci.com/docs/configuration-reference/#docker-execution-environment

**Review by agnivade — CHANGES_REQUESTED** [review]
Blocking the PR temporarily: https://community.mattermost.com/core/pl/s3ujhic5tjn75b18q513sqs7ay

File: build/release.mk
Comment (hmhealey) [reply build/release.mk:154]: Right. That's left over from the old version. It's not actually needed since we're doing a `mkdir -p` when copying the Boards files anyway

File: build/release.mk
Comment (hmhealey) [reply build/release.mk:200]: It's not strictly necessary. I left it there since I thought it might be helpful, but I can remove it

File: .circleci/config.yml
Comment (hmhealey) [reply .circleci/config.yml:83]: I can lower it. I'm not sure what the requirements should be since we're not very consistent with the machine size we use throughout this file (like `check-mocks` uses an xlarge, `check-email-templates` uses a medium+, and `check-store-layers` uses a medium even though they seem like they'd be the same).

**Comment by hmhealey** [comment]
> > but we might also want to consider having it just error out if those files don't exist since I don't know if we intend on having people run make package locally.
> 
> The load tests require devs to run `make package` locally. But even then, we'd want to package the focalboard files as well. So I'd lean towards throwing an error.

I changed that to throw an error and updated the log message with how to fix that.


> Tested out the steps and works fine. The only comment from my side would be to preserve the "make build" workflow for the client side.
> 
> Since I majorly work on the server, I rarely run the webapp. What I do is just do a `make build` on the webapp once, generate the files and keep them there. And just do the dev flow on the server.
> 
> With this change, I need to compile webapp, compile focalboard, AND the key part is to move over the focalboard assets to the right directory. We are doing this step only during the tarball creation job, but it would be beneficial to do it in the webapp `make build` phase itself.
> 
> But it's not a huge blocker since it's a custom workflow that I guess very few developers use and I would be fine in doing 2 more steps manually. I'll leave it to you.

I'll think if there's a good way to make that work properly again. Since production handles product files fundamentally differently than development, it'd be tricky to have that work as well. At the very least, I'm pretty sure your flow will still work for the rest of the web app even if it can't load the products currently.

Unfortunately, that's also one of the things that would be easier if the product files were in `products` instead of `client/products` since products and the web app could have equivalent `make build` targets that write to their respective folders, but that's not something I think I want to change back now 😛 

> Blocking the PR temporarily: https://community.mattermost.com/core/pl/s3ujhic5tjn75b18q513sqs7ay

This has been resolved in https://github.com/mattermost/mattermost-webapp/pull/11355/commits/7cfb035d5ada5726929a2e41547e2373fbd55243

**Comment by phoinixgrr** [comment]
Tests are failing
```
cp: cannot stat '../focalboard/mattermost-plugin/webapp/dist/*': No such file or directory
make[1]: *** [build/release.mk:161: package-general] Error 1
make[1]: Leaving directory '/go/src/github.com/mattermost/mattermost-server'
make: *** [build/release.mk:205: package-osx-amd64] Error 2
```


https://github.com/mattermost/mattermost-server/pull/21401/files#diff-553b9a740f0151f2e959512198dddbdd0d69c398d29583ca0a4e011d5598932eR198

**Comment by agnivade** [comment]
@hmhealey - I am seeing this while trying to load the PRs:

```
Error loading and initializing product boards ChunkLoadError: Loading chunk 72 failed.
(error: https://localhost:8065/static/products/boards/72.js)
while loading "." from 83358
    at v.f.j (remote_entry.js:1:7443)
    at remote_entry.js:1:1047
    at Array.reduce (<anonymous>)
    at v.e (remote_entry.js:1:1025)
    at . (remote_entry.js:1:121)
    at a (remote_entry.js:1:281)
    at i (remotes loading:41:20)
    at c (remotes loading:53:49)
    at i (remotes loading:46:14)
    at remotes loading:52:57
(anonymous) @ products.ts:70
```

**Comment by agnivade** [comment]
But the case where the ff is disabled works fine. It does not load the product, and enabling the plugin works fine. Only the case where ff is enabled fails.

**Comment by hmhealey** [comment]
I think the EE tests are failing because of the error message I added when Boards files don't exist (which is the case for EE builds at the moment).

I'm not sure why that's failing for you, Agniva, but I'll push some more logging improvements there in a moment.

**Review by agnivade — APPROVED** [review]
Verified and looks good!

File: .circleci/config.yml
Comment (spirosoik) [inline .circleci/config.yml:93]: nit: same here

**Review by spirosoik — APPROVED** [review]
lgtm

**Comment by jgilliam17** [comment]
@hmhealey 
[e2e ](https://mattermost-cypress-report.s3.amazonaws.com/252682-0ce76a8-onprem-ent-server-pr-21401-mm-ee-test:0ce76a8/mochawesome.html) without the flag looks good, no PR related failures 
[e2e cycle](https://automation-dashboard.vercel.app/cycles/3b3c60f5-d22b-4cd2-8994-10cdce454e42) failed when flag was set to true
I am still seeing a white screen after trying to share any board.

**Comment by hmhealey** [comment]
Turns out those were all failing because of the default config that Cypress used which had the Boards plugin enabled. The first step of each test was to apply that config, and because the server doesn't let you turn on the plugin and product at the same time, it was erroring out.

I solved that by changing it so that the default E2E test config would have the BoardsProduct flag enabled and the plugin disabled. **That'll be the default in the future once these PRs go in.**

And yes, because of this, me calling the E2E tests with the flag set explicitly was entirely unnecessary

**Review by jgilliam17 — APPROVED** [review]
Tested, e2es - see webapp PR for report.
Looks good to merge

**Comment by phoinixgrr** [comment]
Hello @hmhealey  

Changes introduced here, brake private [bundle](https://git.internal.mattermost.com/ci/bundle) pipelines with: 
```
Unable to find files for Boards product. Please ensure that the Focalboard repository is checked out alongside the server and run 'make build-product' in it.
make[1]: *** [build/release.mk:161: package-general] Error 1
make[1]: Leaving directory '/builds/ci/bundle/mattermost-server'
make: *** [build/release.mk:205: package-osx-amd64] Error 2
```
https://git.internal.mattermost.com/ci/bundle/-/jobs/1154730
https://git.internal.mattermost.com/ci/bundle/-/jobs/1157510


Reported as "failing v7.5 release branch" under https://mattermost.atlassian.net/browse/CLD-4424 by @amyblais 

Kindly advise 
Thanks

**Comment by agnivade** [comment]
Hey @phoinixgrr - What we need to do is add another stage with `onprem-ee-fetch-focalboard` to have the focalboard repo cloned along side with it. Once it has that, the Makefile should be able to compile everything.

**Comment by spirosoik** [comment]
> Hey @phoinixgrr - What we need to do is add another stage with `onprem-ee-fetch-focalboard` to have the focalboard repo cloned along side with it. Once it has that, the Makefile should be able to compile everything.

I will have a quick look too guys.

File: build/release.mk
Comment (edannenberg) [inline build/release.mk:200]: Is it intended that the build fails now if the boards files aren't around? I just want to build and package the mattermost server without the boards stuff as before.

Comment (hmhealey) [reply build/release.mk:200]: Yeah, that was intended, but we could always consider changing it. We're in the process of integrating Boards more into Mattermost instead of having it run as a plugin, and while that's behind a feature flag when running MM in 7.5, it'll eventually become the only way to run Boards in the next few months or so, so it would at least be recommended to have at this point in the packaging scripts. The main reason it errors out is so that we wouldn't accidentally forget those files when packaging our builds, and I did it here since I figured we were the only one using this particular Makefile. I take it you're packaging a custom build using these as well though?

That said, we've also been hoping on not requiring those for development builds at least, so we could possibly make that work for custom production builds as well.

I'll bring this up for discussion next week to see how other people on the team would feel about doing that since  we haven't firmly committed yet to whether or not things like Boards will be a required part of everything in the future since they are likely going to become more integrated with the rest of Mattermost over time.

Comment (hmhealey) [reply build/release.mk:200]: I understand about the concerns of bloat. The idea of it being mandatory has been thrown around once or twice, but it's definitely not something we've committed to. There's people on the team who wouldn't want that to happen either, so you're not alone in that.

As for patching that, currently, there'd be no repercussions for doing that, but in a release or two when we turn the feature on by default (and then on fully) in production builds, the web app will start printing error messages to the console since it'll be trying to load the missing files.

If you're also also building the web app from source with that, maybe we can add an environment variable for both these that'll disable trying to compile the Boards files for the web app? That'd save you from having to patch anything, and it'd still include everything by default like we want on our end. Plus, since we'd have code referencing that on our end, it'd be much less likely that we accidentally remove it without having a good reason to.

**Comment by edannenberg** [comment]
> If you're also also building the web app from source with that, maybe we can add an environment variable for both these that'll disable trying to compile the Boards files for the web app?

Yep the ebuild builds the webapp and the server. A ENV switch would be ideal as Gentoo's package manager has a feature called `use-flags` which are basically feature flags. Could easily add a use-flag that pulls in the boards functionality for people who want it.

File: build/release.mk
Comment (edannenberg) [reply build/release.mk:200]: > If you're also also building the web app from source with that, maybe we can add an environment variable for both these that'll disable trying to compile the Boards files for the web app?

Yep the ebuild builds the webapp and the server. A ENV switch would be ideal as Gentoo's package manager has a feature called `use-flags` which are basically feature flags. Could easily add a use-flag that pulls in the boards functionality for people who want it.

Comment (ThiefMaster) [reply build/release.mk:200]: +1 no keeping Boards (or Playbooks) optional. I'm using Mattermost purely as a communication tool and want to avoid feature creep.

**Pull request merged**
